### DISCUSSION TITLE
How to embedd HTML page under one menu item

### DISCUSSION THREAD

**Discussion by mpusz** [squidfunk/mkdocs-material#6503] [Q&A]
Hi,

I would like to "paste" a HTML page under one position in the MkDocs menu (e.g. Doxygen generated docs). Still, it should have my theme header and footer. I tried with the `template` approach like the main page in your documentation, but I was not able to provide the entire HTML file this way (including all the stylesheets). What is the easiest way to embed such externally provided HTML in the generated docs?

**Comment by alexvoss** [comment]
Hi, can you describe a bit more what you tried and what did not work? It should be possible to include any HTML and any CSS. Whether the styling or interactivity provided by another tool can be made to work with that of Material for MkDocs is another question.

**Comment by mpusz** [comment]
@alexvoss, thanks for your answer.

To be precise, I would like this page https://mpusz.github.io/mp-units/latest/library_reference/core_library of my documentation to embed an HTML document similar to this one https://github.com/JohelEGP/jegp/blob/gh-pages/full.html (here is the rendering https://johelegp.github.io/jegp/full). I would like this page to be rendered only in the page contents part of the window, leaving the top, bottom, and sidebars with the default theme.

**Reply by alexvoss** [reply]
Am not at my PC but as a first stab at the problem, I would suggest looking at the MkDocs hooks mechanism and modifying your page's HTML once the Markdown had been rendered. Parse the content you want to insert with Beautifulsoup and then 'cut-n-paste' content from the result into your document. 

Not the easiest thing to do but what you need is fairly specific, so I would be surprised if there was a shrink wrapped solution.

**Reply by mpusz** [reply]
Ahh, this seems quite complicated for a low-level C++ developer 😉 Hopefully, there is some simpler solution...

**Reply by alexvoss** [reply]
Just to clarify: you can include raw HTML in your Markdown files and with the [Snippets function of the Python Markdown Extensions](https://facelessuser.github.io/pymdown-extensions/extensions/snippets/) you can include files or parts of files in the Markdown before it gets processed. That should make it possible to knock up a prototype...

Right, I tried it but it turns out that the doxygen output needs a little tweaking so that we can cut out the HTML content of body and nothing else. Perhaps this is solvable with a Doxygen configuration or a small `sed` script? Another question is to make sure that the line numbers included are correct but I think that could also be scripted.

Have a look at the [doxygen_meets_mkdocs.zip](https://github.com/squidfunk/mkdocs-material/files/13668332/doxygen_meets_mkdocs.zip) file attached. It contains the tweaked doxygen output and an `index.md` file with a line to include the body content. Hope this helps.

Btw. a proper integration of Doxygen into MkDocs and Material for MkDocs would be a wonderful thing to have. There has been some discussion about this before: https://github.com/mkdocs/mkdocs/discussions/3247 Happy to have a discussion about this. Also, I don't think https://github.com/pieterdavid/mkdocs-doxygen-plugin does what you want, but thought I'd point out it exists.

**Reply by mpusz** [reply]
Thanks a lot for your help! I would not think about such a simple solution 👍🏻 

To be precise, I am not going to provide exactly the docs generated by the Doxygen but by another custom tool developed by @johelegp, which looks like standardeese. This is what I provided here https://johelegp.github.io/jegp/full.

But this should not matter here. The problem remains the same: "How to embed another, possibly generated, HTML file as a subpage of the mkdocs documentation."

**Reply by alexvoss** [reply]
Right, I misunderstood about Doxygen. Does what I suggested solve your problem? If it is 'standardeese' then does the HTML change slowly and you can get away with hand-editing it or a with a bit of `sed`? 

Either way, I am going to shamelessly make use of the fact that I have your attention and ask you if you think that there should be something like [Breathe](https://breathe.readthedocs.io/en/latest/) for (Material for) MkDocs? (I have not come across something like it, but have not exhaustively researched)

> Breathe provides a bridge between the Sphinx and Doxygen documentation systems. It is an easy way to include Doxygen information in a set of documentation generated by Sphinx. The aim is to produce an autodoc like support for people who enjoy using Sphinx but work with languages other than Python. The system relies on the Doxygen’s xml output.

Doxygen makes it easy to write this because of the XML output. I am over 50, so comfortable with the idea of using this ;o)

**Reply by mpusz** [reply]
I was using Sphinx + Breathe + Doxygen for a few years for my project, but it never worked for me as those still did not understand C++20 syntax and were reporting a huge number of parsing issues.  This is why I prefer to write the docs by hand. At least for a modern C++ project.

**Reply by mpusz** [reply]
In our solution, we will write the API reference part by hand and will not depend on the tool to parse it. The output is an HTML with a specific style, and I am still looking for the best way to present it as a subpage of mkdocs. Your solution is enough to obtain the contents, but it does not present proper styles for it.

**Reply by mpusz** [reply]
> I am over 50, so comfortable with the idea of using this

Yeah, I am 45 so I somehow know what you mean 😉

**Reply by squidfunk** [reply]
There's also [MkDoxy](https://github.com/JakubAndrysek/MkDoxy)

**Comment by JohelEGP** [comment]
Try HTML5's `<embed>`.
At <https://www.w3schools.com/tags/tag_embed.asp>, there's an
> Example
> An embedded HTML page:

I also remember HTML4 having something similar.

**Reply by kamilkrzyskow** [reply]
You probably mean the iframe tag which seems to be targeted for embedding websites. However, this doesn't seem to be the solution here as the OP asked for the style to be synced 🤔

Edit: Actually wait, OP only wants the header and footer to stay the same, not convert the style, so perhaps you're onto something.
